lib/luacrypto_aead.c

```diff
@@ -0,0 +1,391 @@
+/*
+* SPDX-FileCopyrightText: (c) 2025 jperon <cataclop@hotmail.com>
+* SPDX-License-Identifier: MIT OR GPL-2.0-only
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+
+/***
+Low-level Lua interface to the Linux Kernel Crypto API for AEAD
+(Authenticated Encryption with Associated Data) ciphers.
+ *
+This module provides a `new` function to create AEAD transform objects,
+which can then be used for encryption and decryption.
+@see crypto_aead_tfm
+
+@module crypto_aead
+ */
+
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/crypto.h>
+#include <crypto/aead.h>
+#include <linux/err.h>
+#include <linux/scatterlist.h>
+#include <linux/slab.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+static const lunatik_class_t luacrypto_aead_tfm_class;
+
+typedef struct {
+	struct crypto_aead *tfm;
+	struct aead_request *req;
+
+	u8 *iv_data;
+	size_t iv_len;
+
+	u8 *work_buffer;
+	size_t work_buffer_len;
+
+	size_t aad_len_in_buffer;
+	size_t crypt_data_len_in_buffer; // For encrypt: PT len. For decrypt: CT+Tag len.
+
+	struct scatterlist sg_work;
+} luacrypto_aead_tfm_t;
+
+LUNATIK_PRIVATECHECKER(luacrypto_check_aead_tfm, luacrypto_aead_tfm_t *);
+
+static void luacrypto_aead_tfm_release(void *private)
+{
+	luacrypto_aead_tfm_t *tfm_ud = (luacrypto_aead_tfm_t *)private;
+	if (!tfm_ud) {
+		return;
+	}
+
+	kfree(tfm_ud->iv_data);
+	tfm_ud->iv_data = NULL;
+	kfree(tfm_ud->work_buffer);
+	tfm_ud->work_buffer = NULL;
+
+	if (tfm_ud->req) {
+		aead_request_free(tfm_ud->req);
+		tfm_ud->req = NULL;
+	}
+
+	if (tfm_ud->tfm && !IS_ERR(tfm_ud->tfm)) {
+		crypto_free_aead(tfm_ud->tfm);
+	}
+}
+
+
+/// AEAD Transform (TFM) methods.
+// These methods are available on AEAD TFM objects created by `crypto_aead.new()`.
+// @see crypto_aead.new
+// @type crypto_aead_tfm
+
+/***
+Sets the encryption key for the AEAD transform.
+@function crypto_aead_tfm:setkey
+@tparam string key The encryption key.
+@raise Error if setting the key fails (e.g., invalid key length for the algorithm).
+ */
```

Comment:
```suggestion
/*
* Sets the encryption key for the AEAD transform.
* @function crypto_aead_tfm:setkey
* @tparam string key The encryption key.
* @raise Error if setting the key fails (e.g., invalid key length for the algorithm).
*/
```

I still have mixed feelings regarding adding this sort of comment.. I'm not sure it belongs here or in the doc.. however, if it belongs here, we should use it as a standard in the framework and agree on the format..

Reply:
It’s from [LDoc](https://stevedonovan.github.io/ldoc/manual/doc.md.html) too, which requires the `/***` (or `///`) to recognize a "doc comment". The advantage is that ldoc is able to extract an api doc from there, as the one of [luaposix](https://luaposix.github.io/luaposix/) for example. Maintaining that at this place has the advantage that, if anyone changes the api, he only has to update the comment to get documentation up-to-date.